File: static/state.js
class Queue {
    constructor() {
        this.pq = [];
    }

    push(item) {
        this.add(item);
    }

    add(item) {
        this.pq.push(item);
        heapSort(this.pq);
    }

    pop() {
        return this.pq.pop()
    }

    peek() {
        return this.pq[0]
    }

    remove(item) {
        if (index_of(this.pq, item) !== -1) {
            this.pq.remove(item);
            heapSort(this.pq);
            return true;
        }
        return false;
    }

    get length() {
        return this.pq.length
    }
}

class State {
    constructor(initial_state) {
        this.steps = [];
        this.parent = null;
        this.depth = 0;
        this.array = clone(initial_state);
    }

    compare(other) {
        return this.score - other.score;
    }

    equals(other) {
        try {
            return this.array === other.array;
        } catch (e) {
            return false;
        }
    }

    lessThan(other) {
        return this.score < other.score;
    }

    contains(item) {
        let pos = index_of(this.array, item);
        try {
            return pos >= 0;
        } catch (e) {
            return false;
        }
    }

    toString() {
        let res = "=========\n";
        for (let i = 0; i < 3; i++) {
            res += "[ ";
            res += this.toMatrix()[i].join(' ').replace('0', 'x');
            res += " ]";
            res += "\r\n";
        }
        res += "=========";
        return res;
    }

    toMatrix() {
        let val = 0;
        let mat = [];
        for (let i = 0; i < 3; i++) {
            let row = [];
            for (let j = 0; j < 3; j++) {
                row[j] = this.array[val];
                val += 1;
            }
            mat[i] = row;
        }
        return mat;
    }

    getPath() {
        let path = [this];
        let pathIndex = 1;
        let state = this.parent;
        while (state.parent) {
            path[pathIndex] = state;
            state = state.parent;
            pathIndex += 1;
        }
        return path.reverse();
    }

    get score() {
        return this.errorsCount() + this.depth;
    };

    errorsCount() {
        let errors = 0;
        for (let i = 0; i < this.array.length; i++) {
            if (this.array[i] !== target[i]) {
                errors += 1;
            }
        }
        return errors;
    }

    get isSolved() {
        return this.errorsCount() <= 0;
    }

    move(direction) {
        let clon = new State(this.array);
        let space_index = index_of(clon.array, 0);

        let row = 0;
        if (space_index < 3) {
            row = 0;
        } else if (space_index >= 6) {
            row = 2;
        } else {
            row = 1;
        }
        let col = 0;
        if (row === 0) {
            col = space_index;
        } else {
            col = space_index - (3 * row);
        }

        let new_row = row;
        let new_col = col;
        let moved = false;

        switch (direction) {
            case 'up':
                if (row > 0) {
                    new_row = row - 1;
                    moved = true;
                }
                break;
            case 'do':
                if (row < 2) {
                    new_row = row + 1;
                    moved = true;
                }
                break;
            case 'le':
                if (col > 0) {
                    new_col = col - 1;
                    moved = true;
                }
                break;
            case 'ri':
                if (col < 2) {
                    new_col = col + 1;
                    moved = true;
                }
                break;
        }

        if (!moved) {
            return {
                moved: false,
                state: clon
            };
        }

        let new_index = space_index;
        if (new_row === 0) {
            new_index = new_col;
        } else {
            new_index = (3 * new_row) + new_col;
        }

        clon.array[space_index] = clon.array[new_index];
        clon.array[new_index] = 0;

        clon.parent = this;
        clon.depth = this.depth + 1;

        let new_steps = [];
        this.steps.forEach((e) => new_steps.push(e));
        new_steps.push(direction.substr(0, 1));
        clon.steps = new_steps;

        return {
            moved: true,
            state: clon
        };
    }

    get getPossibleMovements() {
        let up = this.move('up');
        let down = this.move('do');
        let le = this.move('le');
        let ri = this.move('ri');

        let options = [];
        if (up.moved) {
            options.push(up.state);
        }
        if (down.moved) {
            options.push(down.state);
        }
        if (le.moved) {
            options.push(le.state);
        }
        if (ri.moved) {
            options.push(ri.state);
        }

        timsort.sort(options, (a, b) => a.score - b.score);

        return options;
    }

    createSolutionTree(path) {
        if (this.parent === null) {
            return path;
        } else {
            path.push(this);
            return this.parent.createSolutionTree(path);
        }
    }

    _createNewState(array, parent, depth, steps) {
        let clon = new State(array);
        if (parent !== null) {
            clon.parent = this._createNewState(parent.array,
                                               parent.parent,
                                               parent.depth,
                                               parent.steps);
        } else {
            clon.parent = null
        }
        clon.depth = depth;
        clon.steps = steps;
        return clon;
    }

    _createNewStateObject(array, parent, depth, steps, score) {
        let correctParent = null;
        if (parent !== null) {
            correctParent = this._createNewStateObject(parent.array,
                                                       parent.parent,
                                                       parent.depth,
                                                       parent.steps,
                                                       parent.score);
        } else {
            correctParent = null
        }

        return {
            array: array,
            parent: correctParent,
            depth: depth,
            steps: steps,
            score: score
        };
    }

    solve(print = true, fun = null) {
        let openset = new Queue();

        openset.push(
            this._createNewStateObject(this.array, this.parent, this.depth, this.steps, this.score)
        );

        let closedset = [];
        let moves = 0;
        let iters = 0;

        console.log("Solving..");
        console.log(openset.peek());

        let start = new Date().getTime();
        let end = 0;
        let solved_path = [];

        let pp = false;
        let printed = false;

        while (openset) {
            let current = openset.pop();

            if (iters >= 10000) {
                moves = -1;
                break;
            }

            if (iters % 1000 === 0) {
                if (print) {
                    console.log("Iteration #" + iters);
                }
            }

            if (current.isSolved) {
                end = new Date().getTime();
                let path = current.createSolutionTree([]);
                let reversedPath = clone(path).reverse();

                for (let i = 0; i < reversedPath.length; i++) {
                    if (!printed) {
                        console.log(reversedPath[i]);
                        printed = true;
                    }
                    solved_path.append(reversedPath[i].array);
                    if (fun !== null) {
                        fun(reversedPath[i].array);
                    }
                    if (print) {
                        console.log(reversedPath[i]);
                    }
                }

                moves = path.length;
                let last = path.pop(0);
                if (print) {
                    console.log("Solution found");
                    console.log(moves + " steps: " + last.steps)
                }
                break;
            }

            let obj = this._createNewState(current.array, current.parent, current.depth,
                                           current.steps);
            let new_moves = obj.getPossibleMovements;
            if (new_moves.length > 0) {
                iters += 1;

                for (let j = 0; j < new_moves.length; j++) {
                    let new_state = new_moves[j];
                    let index = index_of(closedset, new_moves[j]);
                    if (index === -1) {
                        openset.push(this._createNewStateObject(new_state.array,
                                                                new_state.parent,
                                                                new_state.depth,
                                                                new_state.steps,
                                                                new_state.score));
                    }
                }
            }
            closedset.push(current);
        }

        if (moves <= 0) {
            console.log("Solution not found!");
        }

        return {
            moves: moves,
            time: end - start,
            solution: solved_path
        };
    }

}

File: State.py
from heapq import heappush, heappop, heapify
from copy import deepcopy
import time

GOAL = [1, 2, 3, 4, 5, 6, 7, 8, 0]
# GOAL = [1, 2, 3, 8, 0, 4, 7, 6, 5]
MAX_ITERS = 5000

UP_KEY = "w"
DOWN_KEY = "s"
LEFT_KEY = "a"
RIGHT_KEY = "d"

UP = "u"
DOWN = "d"
LEFT = "l"
RIGHT = "r"


class Queue:
    def __init__(self):
        self.pq = []

    def append(self, item):
        self.add(item)

    def add(self, item):
        heappush(self.pq, item)

    def pop(self):
        return heappop(self.pq)

    def peek(self):
        return self.pq[0]

    def remove(self, item):
        if item in self.pq:
            copy = deepcopy(item)
            self.pq.remove(item)
            heapify(self.pq)
            return copy is not None
        return False

    def __len__(self):
        return len(self.pq)


# noinspection PyBroadException
class State(object):
    def __init__(self, initial_values):
        self.steps = []
        self.parent = None
        self.depth = 0
        self.mvmnt = ''
        val = 0
        mat = []
        for i in range(0, 3):
            row = []
            for j in range(0, 3):
                row.append(initial_values[val])
                val += 1
            mat.append(row)
        self.matrix = mat

    def __cmp__(self, other):
        return self.matrix == other.matrix

    def __eq__(self, other):
        if self.__class__ != other.__class__:
            return False
        else:
            return self.__cmp__(other)

    def __hash__(self):
        return hash(str(self.matrix))

    def __lt__(self, other):
        return self.score() < other.score()

    def __contains__(self, item):
        pos = self.index_of(item)
        try:
            return pos[0] >= 0 and pos[1] >= 0
        except Exception:
            return False

    def __str__(self):
        res = "=========\n"
        for row in range(3):
            res += "[ "
            res += " ".join(map(str, self.matrix[row])).replace("0", "x")
            res += " ]"
            res += "\r\n"
        res += "========="
        return res

    def clone(self):
        clon = State(self.to_array())
        for i in range(3):
            clon.matrix[i] = self.matrix[i][:]
        return clon

    def get_path(self):
        path = [self]
        state = self.parent
        while state.parent:
            path.append(state)
            state = state.parent
        return reversed(path)

    def score(self):
        return self.errors_count() + self.depth

    def errors_count(self):
        vals = self.to_array()
        errors = 0
        for k in range(len(vals)):
            if vals[k] != GOAL[k]:
                errors += 1
        return errors

    def is_solved(self):
        return self.errors_count() <= 0

    def index_of(self, n):
        ist = [-1, -1]
        for i in range(0, 3):
            for j in range(0, 3):
                try:
                    if self.matrix[i][j] == n:
                        ist = [i, j]
                        break
                except Exception:
                    pass
        return ist

    def to_array(self):
        nums = []
        for i in range(0, 3):
            for j in range(0, 3):
                nums.append(self.matrix[i][j])
        return nums

    def move(self, key):
        if key == UP_KEY:
            return self.move_up()
        elif key == DOWN_KEY:
            return self.move_down()
        elif key == LEFT_KEY:
            return self.move_left()
        elif key == RIGHT_KEY:
            return self.move_right()
        else:
            return False, None

    def move_up(self):
        index = self.index_of(0)
        col = index[1]
        cur_row = index[0]
        new_row = index[0] - 1
        moved = False
        clon = self.clone()
        if 0 <= new_row <= 2:
            pre_val = clon.matrix[new_row][col]
            clon.matrix[new_row][col] = 0
            clon.matrix[cur_row][col] = pre_val

            new_steps = []
            for step in self.steps:
                new_steps.append(step)
            new_steps.append(UP_KEY)
            clon.steps = new_steps

            clon.mvmnt = UP_KEY
            clon.parent = self
            clon.depth = self.depth + 1
            moved = True
        return moved, clon

    def move_down(self):
        index = self.index_of(0)
        col = index[1]
        cur_row = index[0]
        new_row = index[0] + 1
        moved = False
        clon = self.clone()
        if 0 <= new_row <= 2:
            pre_val = clon.matrix[new_row][col]
            clon.matrix[new_row][col] = 0
            clon.matrix[cur_row][col] = pre_val

            new_steps = []
            for step in self.steps:
                new_steps.append(step)
            new_steps.append(DOWN_KEY)
            clon.steps = new_steps

            clon.mvmnt = DOWN_KEY
            clon.parent = self
            clon.depth = self.depth + 1
            moved = True
        return moved, clon

    def move_left(self):
        index = self.index_of(0)
        row = index[0]
        cur_col = index[1]
        new_col = index[1] - 1
        moved = False
        clon = self.clone()
        if 0 <= new_col <= 2:
            pre_val = clon.matrix[row][new_col]
            clon.matrix[row][new_col] = 0
            clon.matrix[row][cur_col] = pre_val

            new_steps = []
            for step in self.steps:
                new_steps.append(step)
            new_steps.append(LEFT_KEY)
            clon.steps = new_steps

            clon.mvmnt = LEFT_KEY
            clon.parent = self
            clon.depth = self.depth + 1
            moved = True
        return moved, clon

    def move_right(self):
        index = self.index_of(0)
        row = index[0]
        cur_col = index[1]
        new_col = index[1] + 1
        moved = False
        clon = self.clone()
        if 0 <= new_col <= 2:
            pre_val = clon.matrix[row][new_col]
            clon.matrix[row][new_col] = 0
            clon.matrix[row][cur_col] = pre_val

            new_steps = []
            for step in self.steps:
                new_steps.append(step)
            new_steps.append(RIGHT_KEY)
            clon.steps = new_steps

            clon.mvmnt = RIGHT_KEY
            clon.parent = self
            clon.depth = self.depth + 1
            moved = True
        return moved, clon

    def get_possible_movements(self):
        mu, att_one = self.move_up()
        md, att_two = self.move_down()
        ml, att_thr = self.move_left()
        mr, att_fou = self.move_right()

        options = []
        if mu:
            options.append(att_one)
        if md:
            options.append(att_two)
        if ml:
            options.append(att_thr)
        if mr:
            options.append(att_fou)

        return sorted(options, key=lambda p: p.score())

    def create_solution_tree(self, path):
        if self.parent is None:
            return path
        else:
            path.append(self)
            return self.parent.create_solution_tree(path)

    def _a_star(self, max_duration=120):
        openset = Queue()
        openset.add(self.clone())
        closedset = []
        start = time.time()

        while openset:
            end = time.time()
            if float(end - start) > max_duration or len(openset) <= 0:
                return None, -1

            current = openset.pop()
            if current.is_solved():
                moves = []
                temp = current
                while True:
                    moves.insert(0, temp.mvmnt)
                    if temp.depth <= 1:
                        break
                    temp = temp.parent
                end = time.time()
                return moves, float(end - start)
            new_moves = current.get_possible_movements()
            for state in new_moves:
                if state not in closedset:
                    openset.add(state)
            closedset.append(current)
        return None, -1

    def _bfs(self, max_duration=120):
        start = time.time()
        nodes = [self.clone()]
        while nodes:
            end = time.time()
            if float(end - start) > max_duration or len(nodes) <= 0:
                return None, -1

            node = nodes.pop(0)
            if node.is_solved():
                moves = []
                temp = node
                while True:
                    moves.insert(0, temp.mvmnt)
                    if temp.depth <= 1:
                        break
                    temp = temp.parent
                end = time.time()
                return moves, float(end - start)

            nodes.extend(node.get_possible_movements())
        return None, -1

    def _dfs(self, max_depth=20, max_duration=120):
        start = time.time()
        nodes = [self.clone()]
        while nodes:
            end = time.time()
            if float(end - start) > max_duration or len(nodes) <= 0:
                return None, -1

            node = nodes.pop(0)
            if node.is_solved():
                moves = []
                temp = node
                while True:
                    moves.insert(0, temp.mvmnt)
                    if temp.depth <= 1:
                        break
                    temp = temp.parent
                end = time.time()
                return moves, float(end - start)
            # Add all the expansions to the beginning of the stack if we are under the depth limit
            if node.depth < max_depth:
                new_nodes = node.get_possible_movements()
                new_nodes.extend(nodes)
                nodes = new_nodes
        return None, -1

    def _ids(self, max_depth=10, max_duration=120):
        start = time.time()
        for i in range(max_depth):
            end = time.time()
            if float(end - start) > max_duration:
                return None, -1
            result, total = self._dfs(max_depth=i, max_duration=max_duration)
            if result is not None:
                return result, total
        return None, -1

    def optimal_solution(self, max_duration=30, include_ids=False, use_all=False):
        print("Attempting to solve:")
        print(str(self))
        print("... ...")

        algo = "A*"

        print("Attempting to use %s algorithm" % algo)
        path, total = self._a_star(max_duration=max_duration * (1 if use_all else 1.5))

        if use_all:
            if path is None or len(path) <= 0:
                algo = "Breadth First Search"
                print("Attempting to use %s algorithm" % algo)
                path, total = self._bfs(max_duration=max_duration - 5)
            if path is None or len(path) <= 0:
                algo = "Depth First Search"
                print("Attempting to use %s algorithm" % algo)
                path, total = self._dfs(max_duration=max_duration - 5)
            if path is None or len(path) <= 0:
                if include_ids:
                    algo = "Iterative Depth First Search"
                    print("Attempting to use %s algorithm" % algo)
                    path, total = self._ids(max_duration=max_duration)

        if path is None or len(path) <= 0:
            algo = "None"
            path = []
            total = -1

        if path is not None and len(path) > 0 and total >= 0:
            print("Solution found using %s algorithm in %d movement(s) and %.3f seconds" % (
                algo, len(path), total))
            print("Steps: " + str(path))
        else:
            print("Solution not found")
        return path, total
